src/Client/HedgedConnections.cpp

```diff
@@ -0,0 +1,500 @@
+{
+    /// When we receive first packet of data from replica, we stop working with replicas, that are
+    /// responsible for the same offset.
+    OffsetState & offset_state = offset_states[replica_location.offset];
+    offset_state.replicas[replica_location.index].change_replica_timeout.reset();
+    ++offsets_with_received_first_data_packet;
+    offset_state.first_packet_of_data_received = true;
+
+    for (size_t i = 0; i != offset_state.replicas.size(); ++i)
+    {
+        if (i != replica_location.index && offset_state.replicas[i].connection)
+        {
+            offset_state.replicas[i].connection->sendCancel();
+            finishProcessReplica(offset_state.replicas[i], true);
+        }
+    }
+
+    /// If we received data from replicas with all offsets, we need to stop choosing new replicas.
+    if (hedged_connections_factory.hasEventsInProcess() && offsets_with_received_first_data_packet == offset_states.size())
+    {
+        if (next_replica_in_process)
+        {
+            epoll.remove(hedged_connections_factory.getFileDescriptor());
+            next_replica_in_process = false;
+        }
+        hedged_connections_factory.stopChoosingReplicas();
+    }
+}
+
+void HedgedConnections::tryGetNewReplica(bool start_new_connection)
+{
+    Connection * connection = nullptr;
+    HedgedConnectionsFactory::State state = hedged_connections_factory.getNextConnection(start_new_connection, false, connection);
+
+    /// Skip replicas that doesn't support two-level aggregation if we didn't disable it in sendQuery.
+    while (state == HedgedConnectionsFactory::State::READY && !disable_two_level_aggregation
+           && connection->getServerRevision(hedged_connections_factory.getConnectionTimeouts())
+               < DBMS_MIN_REVISION_WITH_CURRENT_AGGREGATION_VARIANT_SELECTION_METHOD)
+        state = hedged_connections_factory.getNextConnection(true, false, connection);
+
+    if (state == HedgedConnectionsFactory::State::READY)
+    {
+        size_t offset = offsets_queue.front();
+        offsets_queue.pop();
+
+        offset_states[offset].replicas.emplace_back(connection);
+        ++offset_states[offset].active_connection_count;
+        offset_states[offset].next_replica_in_process = false;
+        ++active_connection_count;
+
+        ReplicaState & replica = offset_states[offset].replicas.back();
+        epoll.add(replica.packet_receiver.getFileDescriptor());
+        fd_to_replica_location[replica.packet_receiver.getFileDescriptor()] = ReplicaLocation{offset, offset_states[offset].replicas.size() - 1};
+        epoll.add(replica.change_replica_timeout.getDescriptor());
+        timeout_fd_to_replica_location[replica.change_replica_timeout.getDescriptor()] = ReplicaLocation{offset, offset_states[offset].replicas.size() - 1};
+
+        pipeline_for_new_replicas.run(replica);
+    }
+    else if (state == HedgedConnectionsFactory::State::NOT_READY && !next_replica_in_process)
```

Comment:
Now I think that maybe `next_replica_in_process` is actually a property of `HedgedConnectionsFactory`.
It may be not just flag, but the number-of-processing-replicas (e.g. the number of calls `hedged_connections_factory.getNextConnection(true, ...)` minus the number of `READY` states returned).

And invariant may be `number-of-processing-replicas != 0` <=> `epoll contains edged_connections_factory descriptor`.

When we call `hedged_connections_factory.stopChoosingReplicas()`, number-of-processing-replicas is set to 0